master/internal/run/runs.go

```diff
@@ -0,0 +1,178 @@
+package run
+
+import (
+	"fmt"
+	"reflect"
+	"regexp"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/determined-ai/determined/master/pkg/model"
+)
+
+const (
+	integerType   = "integer"
+	floatType     = "float"
+	booleanType   = "boolean"
+	stringType    = "string"
+	timestampType = "timestamp"
+	// MaxMetadataValueStringLength is the maximum length of a metadata value.
+	MaxMetadataValueStringLength = 5000
+	// MaxMetadataKeyLength is the maximum length of a metadata key.
+	MaxMetadataKeyLength = 50
+	// MaxMetadataArrayLength is the maximum length of a metadata array.
+	MaxMetadataArrayLength = 100
+	// MaxMetadataDepth is the maximum depth of nested metadata.
+	MaxMetadataDepth = 10
+	// ExcludedMetadataCharactersPattern is the pattern of characters that are not allowed in metadata values.
+	ExcludedMetadataCharactersPattern = "[\\$\\.\\[\\]]"
+	// MaxKeyCount is the maximum number of metadata keys allowed for a run.
+	MaxKeyCount = 1000
+)
+
+// FlattenMetadata flattens a nested map of run metadata into a list of RunMetadataIndex entries.
+func FlattenMetadata(
+	data map[string]any,
+) (flatMetadata []model.RunMetadataIndex, err error) {
+	if len(data) == 0 {
+		return nil, fmt.Errorf("metadata is empty")
+	}
+	flatMetadata, err = flattenMetadata(data)
+	if err != nil {
+		return nil, fmt.Errorf("error flattening metadata: %w", err)
+	}
+	return flatMetadata, nil
+}
+
+// parseTimestamp converts a string to a timestamp and returns the timestamp.
+func parseTimestamp(value string) (string, error) {
+	formats := []string{
+		time.RFC3339Nano,
+		time.RFC3339,
+		"2006-01-02",
+		"2006-01",
+	}
+	for _, format := range formats {
+		if timestamp, err := time.Parse(format, value); err == nil {
+			return timestamp.UTC().Format(time.RFC3339Nano), nil
+		}
+	}
+	return "", fmt.Errorf("unable to parse timestamp")
+}
+
+// parseMetadataValueType converts a value to a string and returns the type of the value.
+func parseMetadataValueType(rawValue any) (value string, valueType string) {
```

Comment:
summary metrics does this same kinda thing

https://github.com/determined-ai/determined/blob/84299a660fd86417c4bc11d134dd8462597cdd51/master/internal/db/postgres_trial.go#L722

would be cool if we only had one place where we did this, since it is easy to get wrong. I would check to see what happens if you report ```NaN``` in python code if it becomes a float

Reply:
Sounds great! I'm down to improve the reuse!